### PR TITLE
Support move semantics for PinnableSlice

### PR DESCRIPTION
**Pull request by cheng-chang**
Summary:
It's logically correct for PinnableSlice to support move semantics to transfer ownership of the pinned memory region. This PR adds both move constructor and move assignment to PinnableSlice.

Test Plan:
A set of unit tests for the move semantics are added in slice_test.
So `make slice_test && ./slice_test`.

### CODE REVIEW

File: util/slice.cc [hunk 1]
```diff
@@ -209,4 +209,33 @@ const SliceTransform* NewNoopTransform() {
...
+      if (other.buf_ == &other.self_space_) {
+        self_space_ = std::move(other.self_space_);
+        buf_ = &self_space_;
+        data_ = buf_->data();
+      } else {
+        buf_ = other.buf_;
+        data_ = other.data_;
+      }
+    }
+    other.size_ = 0;
```
Comment (anand1976) [inline util/slice.cc]: Would be a good idea to also reset ```other.buf_``` to its ```self_space_```. What would be the value of ```other.data_```?

File: util/slice_test.cc [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    PinnableSlice v2(std::move(v1));
+
+    AssertEmpty(v1);
+    // Since v1's Cleanable has been moved to v2,
+    // no cleanup should happen in Reset.
+    v1.Reset();
+    ASSERT_EQ(1, res);
+
+    AssertSameData(const_str1, v2);
+  }
```
Comment (anand1976) [inline util/slice_test.cc]: Can we try reusing ```v1``` to make sure its in a valid state?

File: util/slice.cc (hunk 1 above)
Comment (cheng-chang) [reply util/slice.cc]: to be consistent, after resetting `other.buf_` to `other.self_space_`, `other.data_` needs to be set to `other.buf_`.

**Comment by cheng-chang** [comment]
@anand1976 PTAL

**Comment by cheng-chang** [comment]
@anand1976 updated the build files, thanks.

**Comment by cheng-chang** [comment]
The failed test is the known flakey env_test.

**Pull request closed without merging**